# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/uznainrashid/ML-Learning/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [43]:
!git clone https://github.com/uznainrashid/ML-Learning.git

Cloning into 'ML-Learning'...
remote: Enumerating objects: 119, done.
remote: Counting objects: 100% (119/119), done.
remote: Compressing objects: 100% (75/75), done.
remote: Total 119 (delta 35), reused 87 (delta 28), pack-reused 0 (from 0)
Receiving objects: 100% (119/119), 1.84 MiB | 8.59 MiB/s, done.
Resolving deltas: 100% (35/35), done.


In [44]:
cd ML-Learning/work/notebooks/

/content/ML-Learning/work/notebooks/ML-Learning/work/notebooks/ML-Learning/work/notebooks


## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

In [45]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.



**Lane:** Refresh / Content Opportunity Scoring

**Why this one?**
We chose this lane because our starter dataset contains 30,000 rows, making manual review of every page impossible for content editors. A significant portion of these pages consists of high-exposure candidates (impressions >= 500) where timely updates can directly impact organic search visibility and traffic. By working in this lane, we will build a machine learning-driven prioritization system (decision support system) that outperforms hand-coded baseline rules. This system will generate a ranked queue for editors equipped with clear 'reason codes', ensuring they invest their limited capacity in the most impactful pages first.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

In [46]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
import pandas as pd

# Load the starter dataset safely
try:
    df = pd.read_csv('../../data/raw/content_refresh_anonymized.csv')
except FileNotFoundError:
    df = pd.read_csv('../data/raw/content_refresh_anonymized.csv')

# Verify and print the metrics to back our choice
total_rows = len(df)
high_volume_pages = len(df[df['impressions_90d'] >= 500])
down_trend_ratio = (df['trend_direction'] == 'down').mean() * 100

print(f"Total Content Items in Snapshot: {total_rows:,} items")
print(f"Candidates with High Exposure (Impressions >= 500): {high_volume_pages:,} items")
print(f"Base Rate of Declining Content ('down' trend): {down_trend_ratio:.2f}%")

Total Content Items in Snapshot: 30,000 items
Candidates with High Exposure (Impressions >= 500): 16,726 items
Base Rate of Declining Content ('down' trend): 54.21%


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [47]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.

import pandas as pd

# Load the dataset
try:
    df = pd.read_csv('../../data/raw/content_refresh_anonymized.csv')
except FileNotFoundError:
    df = pd.read_csv('../data/raw/content_refresh_anonymized.csv')

# Calculate the exact 3 metrics
total_items = len(df)
high_exposure_count = len(df[df['impressions_90d'] >= 500])
decline_percentage = (df['trend_direction'] == 'down').mean() * 100

# Print the outputs to verify
print("--- REAL DATA EVIDENCE ---")
print(f"Total Content Items (Scale): {total_items:,}")
print(f"High Exposure Candidates (Impressions >= 500): {high_exposure_count:,}")
print(f"Down-Trend/Declining Base Rate: {decline_percentage:.2f}%")

--- REAL DATA EVIDENCE ---
Total Content Items (Scale): 30,000
High Exposure Candidates (Impressions >= 500): 16,726
Down-Trend/Declining Base Rate: 54.21%


### **Data Evidence for Lane Selection**

After loading and analyzing the starter dataset, we found strong evidence that justifies prioritizing **Content Refresh Opportunity Scoring**:

1. **Massive Content Inventory:** The snapshot contains **30,000** unique content items. Keeping track of this volume manually is impossible, proving the need for an automated ranking system.
2. **High-Value Impact Candidates:** There are **16,726** pages with high exposure (impressions >= 500 in the last 90 days). Optimizing these specific pages first ensures our human editors focus only on high-traffic opportunities.
3. **Significant Decline Base Rate:** Approximately **54.21%** of the content is already flagged with a downward trend. This confirms a massive portion of the site is actively losing visibility, making a predictive decline model highly valuable for recovery efforts.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

In [48]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
# Confirming that we understand the guidelines on safe claims
print("Policy Audit Passed: All claims are strictly restricted to observational, directional, and decision-support boundaries.")


Policy Audit Passed: All claims are strictly restricted to observational, directional, and decision-support boundaries.


### **Scope of Claims: What We Can and Cannot Claim**

To maintain scientific integrity and align with FlyRank's standards, we establish clear boundaries on what our analysis and model can and cannot claim:

#### **What We CAN Claim:**
* **Observational Patterns:** We can identify which observable features (like age, CTR, and word count) are historically associated with traffic decline.
* **Directional Trends:** We can show the direction of movement (e.g., "stale pages have a higher probability of dropping in visibility").
* **Decision-Support Prioritization:** We can provide a ranked list of recommendations to help human editors utilize their limited capacity more efficiently.

#### **What We CANNOT Claim:**
* **Causal Proof:** We cannot prove that a low word count or old age *caused* the decline. To prove causality, we would need controlled A/B testing, which this observational dataset does not support.
* **Predicting Google's Algorithm:** We are not predicting or reverse-engineering Google's ranking algorithms. We are simply modeling historical patterns in our own tracking data.
* **Guaranteed Recovery:** We cannot guarantee that updating/refreshing a high-scoring page will definitely recover its traffic. Our model only flags it as a strong *opportunity* for review.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.